### PR TITLE
Add log line for when ledger entry cannot be found

### PR DESCRIPTION
**Pull request by jumaffre**
We now log when a ledger entry cannot be found by the `Ledger` host class. This is useful for operators when a user requests historical entries that cannot be found because the historical ledger files have not been copied to the the node's local directory (e.g. when starting a new node from a snapshot).

Also improved and cleanup the test that checks that node started from a snapshot, but without an historical ledger cannot serve historical queries.

### CODE REVIEW

**Review by achamayou — APPROVED** [review]

File: src/host/ledger.h [hunk 1]
```diff
@@ -366,7 +366,11 @@ namespace asynchost
     {
       if ((from < start_idx) || (to > get_last_idx()) || (to < from))
       {
-        LOG_FAIL_FMT("Unknown entries range: {} - {}", from, to);
+        LOG_FAIL_FMT(
+          "Unknown entries range: {} - {} in ledger file {}",
```
Comment (achamayou) [inline src/host/ledger.h:370]: Maybe "Could not find entries {} to {} ..." is a little clearer, or maybe it's just me, I don't know :)

File: tests/reconfiguration.py [hunk 2]
```diff
@@ -217,8 +207,29 @@ def test_add_node_from_snapshot(
...
+            idx=idx,
+            msg=historical_entry["msg"],
+            seqno=historical_entry["seqno"],
+            view=historical_entry["view"],
+            historical=True,
+        )
+    except infra.logging_app.LoggingTxsVerifyException:
+        assert (
+            not copy_ledger
+        ), f"New node {new_node.local_node_id} without ledger shoud not be able to serve historical entries"
```
Comment (eddyashton) [inline tests/reconfiguration.py:228]:
```suggestion
        ), f"New node {new_node.local_node_id} without ledger should not be able to serve historical entries"
```

File: tests/reconfiguration.py [hunk 3]
```diff
@@ -217,8 +207,29 @@ def test_add_node_from_snapshot(
...
+    else:
+        assert (
+            copy_ledger
+        ), f"New node {new_node.local_node_id} with ledger shoud be able to serve historical entries"
```
Comment (eddyashton) [inline tests/reconfiguration.py:232]:
```suggestion
        ), f"New node {new_node.local_node_id} with ledger should be able to serve historical entries"
```

**Review by eddyashton — APPROVED** [review]

File: python/ccf/ledger.py [hunk 4]
```diff
@@ -1001,8 +1001,19 @@ def get_latest_public_state(self) -> Tuple[dict, int]:
...
+                # If a transaction cannot be read (e.g. because it was only partially written to disk
+                # before a crash), return public state so far. This is consistent with CCF's behaviour
+                # which discards the incomplete transaction on recovery.
+                try:
+                    public_domain = tx.get_public_domain()
+                except Exception:
+                    print(
+                        f"Error reading ledger entry. Latest read seqno: {latest_seqno}"
+                    )
+                    return public_tables, latest_seqno
```
Comment (jumaffre) [inline python/ccf/ledger.py:1013]: Also adding some resilience to the Python infra ledger reading that sometimes fails in the new `recovery_benchmark.sh`: https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=45908&view=logs&j=5435e0ac-25e5-5426-50be-61b0d0ea8d34&t=20753e5b-87ba-5bdc-0f17-d34c09f927a7&l=195

**Comment by jumaffre** [comment]
Merging this now that the IceLake CI pool is back.

**Comment by jumaffre** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|main|[](https://github.com/microsoft/CCF/pull/3896)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**